bin/exec.ml

```diff
@@ -32,6 +32,75 @@ let man =
 
 let info = Cmd.info "exec" ~doc ~man
 
+module Command_to_exec = struct
+  (* A command to execute, which knows how to (re)build the program and then
+     run it with some arguments in an enivorment *)
+
+  type t =
+    { get_path_and_build_if_necessary :
+        unit -> (Path.t, [ `Already_reported ]) result Fiber.t
+    ; args : string list
+    ; env : Env.t
+    }
+
+  (* Helper function to spawn a new process running a command in an
+     environment, returning the new process' pid *)
+  let spawn_process path ~args ~env =
+    let path = Path.to_string path in
+    let env = Env.to_unix env |> Spawn.Env.of_list in
+    let argv = path :: args in
+    let pid = Spawn.spawn ~prog:path ~env ~argv () in
+    Pid.of_int pid
+
+  (* Run the command, first (re)building the program which the command is
+     invoking *)
+  let build_and_run_in_child_process
+      { get_path_and_build_if_necessary; args; env } =
+    get_path_and_build_if_necessary ()
+    |> Fiber.map ~f:(Result.map ~f:(spawn_process ~args ~env))
+end
+
+module Watch = struct
+  (* When running `dune exec` in watch mode, this will keep track of the pid of
+     the process created to run the program in the previous iteration so that
+     it can be killed (for long running programs, e.g. servers) and restarted
+     when its source is changed. *)
+
+  type state = { currently_running_pid : Pid.t option ref }
+
+  let init_state () = { currently_running_pid = ref None }
+
+  (* Helper function to kill a process and block the calling thread until the
+     killed process stops. The pid of the killed process will be reaped. *)
+  let kill_and_reap_process pid =
+    let pid_int = Pid.to_int pid in
+    Unix.kill pid_int Sys.sigkill;
```

Comment:
`sigkill` is a bit aggressive. `sigstop` is a bit gentler.

Reply:
do you mean `sigterm`? `sigstop` will "pause" the process.

Reply:
I initially used `sigkill` here because on windows it's the only signal emulated. On unix I think we should send `sigterm` and then if the process is still running after some period, send sigkill. `Scheduler.wait_for_process` waits for a process with a timeout and sends `sigkill` if the timeout expires so that might be helpful, but it looks like it raises `Build_cancelled` if the build is restarted while it's waiting.

Reply:
> I initially used sigkill here because on windows it's the only signal emulated

You can keep using Sygkill on Windows.

> Scheduler.wait_for_process waits for a process with a timeout and sends sigkill if the timeout expires so that might be helpful, but it looks like it raises Build_cancelled if the build is restarted while it's waiting.

The behavior that we would like would be:

1. Send sigterm
2. Wait for a timeout
3. Send sigkill if the process isn't dead

It would of course be nice if we could sure that with the rest of the code somehow.

Reply:
Updated to wait send sigterm and then wait for a timeout before sending sigkill if the process is still running. I implemented this with periodic calls to `wait [ WNOHANG ]` because it seems simpler than a blocking call to `wait` in a separate thread.